product_kit_management/models/sale_order_line.py

```diff
@@ -0,0 +1,64 @@
+from odoo import Command, api, fields, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    is_kit = fields.Boolean(string="Is Kit", related="product_id.is_kit", store=True)
+    kit_component_ids = fields.One2many(
+        'product.kit.component', 'parent_order_line_id',
+        compute="_compute_sub_products", store=True, string="Kit Components",
+        help="List of sub products"
+    )
+    is_kit_component = fields.Boolean(string="Is Kit Component?", default=False)
+    kit_parent_line_id = fields.Many2one(
+        'sale.order.line', string="Parent Kit Line",
+        help="Refrence to the main kit product's sale order line"
+    )
+
+    @api.depends('product_id')
+    def _compute_sub_products(self):
+        """
+        Computes the sub-products (kit components) for a sale order line.
+        It retrieves the sub-products linked to the main product and updates
+        the `kit_component_ids` field accordingly.
+        """
+        for line in self:
+            product_template = line.product_id.product_tmpl_id  
+            sub_products = product_template.sub_products_id  
+            commands = []
+            existing_components = {comp.kit_component_id.id: comp for comp in line.kit_component_ids}
+            for sub_product in sub_products:
+                if sub_product.id in existing_components:
+                    continue
+                else:
+                    commands.append(Command.create({
+                        'parent_order_line_id': line.id,
+                        'kit_component_id': sub_product.id,
+                        'price': sub_product.lst_price or 0.0,
+                    }))
+            line.kit_component_ids = commands
+
+    @api.ondelete(at_uninstall=False)
+    def _delete_sub_products(self):
+        """
+        Ensures that when a kit product is deleted from a sale order,
+        all its linked kit component lines are also removed.
+        """
+        for line in self:
+            if line.is_kit:
+                self.env['sale.order.line'].search([
+                    ('order_id', '=', line.order_id.id),
+                    ('kit_parent_line_id', '=', line.id)
+                ]).unlink()
+
+    def write(self, vals):
+        """
+        Prevents manual modification of key fields (`product_id`, `product_uom_qty`, `price_unit`)
+        for sub-products (components of a kit) to maintain data integrity.
+        """
+        if any(line.is_kit_component for line in self):
+            restricted_fields = {'product_id', 'product_uom_qty', 'price_unit'}
+            if any(field in vals for field in restricted_fields):
+                raise models.ValidationError("Sub-products of a kit cannot be modified manually.")
```

Comment:
```suggestion
        if (
            any(line.is_kit_component for line in self)
            and any(field in vals for field in {'product_id', 'product_uom_qty', 'price_unit'})
        ):
            raise models.ValidationError("Sub-products of a kit cannot be modified manually.")
```
Why from `model.ValidationError`?